src/data/finsupp/basic.lean

```diff
@@ -1150,12 +1155,42 @@ finset.subset.trans support_sum $
   finset.subset.trans (finset.bUnion_mono $ assume a ha, support_single_subset) $
   by rw [finset.bUnion_singleton]; exact subset.refl _
 
+@[simp]
+lemma map_domain_sum_index {X Y : Type*} (f : X → Y)
+  {M N : Type*} [add_comm_monoid M] [add_comm_monoid N]
+  (g : X →₀ M) (h : Y → M →+ N) :
+  (finsupp.map_domain f g).sum (λ y m, h y m) = g.sum (λ x m, h (f x) m) :=
+begin
+  dsimp [finsupp.map_domain],
+  apply finsupp.induction_linear g,
+  { simp, },
+  { intros g₁ g₂ h₁ h₂,
+    rw [finsupp.sum_add_index, finsupp.sum_add_index, finsupp.sum_add_index, h₁, h₂];
+    simp, },
+  { simp, }
+end
+
 @[to_additive]
 lemma prod_map_domain_index [comm_monoid N] {f : α → β} {s : α →₀ M}
-  {h : β → M → N} (h_zero : ∀a, h a 0 = 1) (h_add : ∀a b₁ b₂, h a (b₁ + b₂) = h a b₁ * h a b₂) :
-  (map_domain f s).prod h = s.prod (λa b, h (f a) b) :=
+  {h : β → M → N} (h_zero : ∀b, h b 0 = 1) (h_add : ∀b m₁ m₂, h b (m₁ + m₂) = h b m₁ * h b m₂) :
+  (map_domain f s).prod h = s.prod (λa m, h (f a) m) :=
 (prod_sum_index h_zero h_add).trans $ prod_congr rfl $ λ _ _, prod_single_index (h_zero _)
 
+/--
+A version of `sum_map_domain_index` that takes a bundled `add_monoid_hom`,
+rather than separate linearity hypotheses.
+-/
+-- Note that in `prod_map_domain_index`, `M` is still an additive monoid,
+-- so there is no analogous version in terms of `monoid_hom`.
+@[simp]
```

Comment:
Out of scope for this PR perhaps, but it would be nice to have this also for add_equiv, linear_map, linear_equiv, alg_hom, alg_equiv

Reply:
I think I'll pass, for now. :-)